### PR TITLE
SSL Pinned Certificate

### PR DESCRIPTION
**Pull request by enahum**
#### Summary
Summary
As part of some of the requirements by XXXXX we needed to add support for SSL Pinning when building your own app.

The way is being built is a combination of the app and the network-client library, the app will hold the certificate assets while the network-library will use these certs to verify the server trust.

The cert files should have either a .crt or .cer extension while the name of the file must be the domain name.

The reason we support two files extensions per domain is to be able to build the app taking into account server cert rotation and minimizing the amount of app distribution through updates in order to validate the certs.

This SSL Pinning certificates should be included in the app at build time.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-59055

#### Release Note
```release-note
Added support for SSL Pinning certificates on build your own
```

### CODE REVIEW

**Review by larkox — APPROVED** [review]
LGTM,but a few minor things to verify.

File: ios/Gekidou/Sources/Gekidou/Networking/Networki+Error.swift [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import Foundation
```
Comment (rahimrahman) [inline ios/Gekidou/Sources/Gekidou/Networking/Networki+Error.swift:1]: Maybe the name was accidentally spelled wrong:

`Network+Error.swift` vs `Network(i)+Error.swift`

File: android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.kt [hunk 2]
```diff
@@ -103,8 +103,10 @@ class CustomPushNotification(
 
         when (type) {
             CustomPushNotificationHelper.PUSH_TYPE_MESSAGE, CustomPushNotificationHelper.PUSH_TYPE_SESSION -> {
-                val currentActivityName = mAppLifecycleFacade.runningReactContext?.currentActivity?.componentName?.className ?: ""
-                Log.i("ReactNative", currentActivityName)
+                val currentActivityName = try {
+                    mAppLifecycleFacade.runningReactContext.currentActivity?.componentName?.className ?: ""
+                } catch (e: Exception) { "" }
+
```
[the remark this answers is not in the document]
Comment (enahum) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.kt:109]: my approach is better as the issue is that `runningReactContext` can be null and trying to access `currentActivity` will throw, even if unrelated.. needs fixing.. and sure I'll make the change in another PR if you prefer.

[the remark this answers is not in the document]
Comment (enahum) [reply android/app/src/main/java/com/mattermost/rnbeta/CustomPushNotification.kt]: reverted to your change

**Review by larkox — APPROVED** [review]

**Comment by rahimrahman** [comment]
@enahum as discussed last night, I'm going to excuse myself from being the reviewer for this PR as even after spending time to understand what's going on (with external help), I am not comfortable to approve or offer any suggestions.

File: libraries/@mattermost/rnshare/android/src/main/java/com/mattermost/rnshare/ShareWorker.kt [hunk 3]
```diff
@@ -22,11 +24,60 @@ import org.json.JSONException
...
+    private fun getCertificateFingerPrint(certInputStream: InputStream): String {
+        val certFactory = CertificateFactory.getInstance("X.509")
+        val certificate = certFactory.generateCertificate(certInputStream) as X509Certificate
+        val sha256 = MessageDigest.getInstance("SHA-256")
+        Base64.encodeToString(sha256.digest(certificate.publicKey.encoded), Base64.NO_WRAP)
+        val fingerprintBytes = sha256.digest(certificate.publicKey.encoded)
+        return Base64.encodeToString(fingerprintBytes, Base64.NO_WRAP)
+    }
+
+    private fun getCertificatesFingerPrints(): Map<String, List<String>> {
```
Comment (cpoile) [inline libraries/@mattermost/rnshare/android/src/main/java/com/mattermost/rnshare/ShareWorker.kt:61]: Is there a way to share the code with the network client?

**Review by cpoile — APPROVED** [review]
Nice work, thanks Elias!

File: libraries/@mattermost/rnshare/android/src/main/java/com/mattermost/rnshare/ShareWorker.kt (hunk 3 above)
Comment (enahum) [reply libraries/@mattermost/rnshare/android/src/main/java/com/mattermost/rnshare/ShareWorker.kt:61]: There is if we add the network client as a dependency of this one, so in the end even if the code is dupe is less expensive this way

**Review by esarafianou — COMMENTED** [review]
Similarly, almost done with the review.

File: ios/Gekidou/Sources/Gekidou/Networking/ShareExtension+Post.swift [hunk 4]
```diff
@@ -98,9 +98,11 @@ extension ShareExtension {
                 id
             )
             return
-            
         }
         
+        self.removeUploadSessionData(id: id)
+        self.deleteUploadedFiles(files: data.files)
```
[the remark this answers is not in the document]
Comment (enahum) [reply ios/Gekidou/Sources/Gekidou/Networking/ShareExtension+Post.swift:104]: Cause don't really need to wait for the callback to be executed in prder to do some cleanup, so thid way is more efficient

File: app/managers/network_manager.ts [hunk 5]
```diff
@@ -129,12 +153,23 @@ class NetworkManager {
...
-                    'The certificate for this server is invalid.\nYou might be connecting to a server that is pretending to be “{hostname}” which could put your confidential information at risk.',
-                ).replace('{hostname}', parsed.hostname),
+                this.intl.formatMessage(messages.invalidSslTitle),
+                this.intl.formatMessage(messages.invalidSSLDescription, {hostname: parsed.hostname}),
+            );
+        } else if (SERVER_TRUST_EVALUATION_FAILED === event.errorCode && !showingServerTrustAlert) {
+            logDebug('Invalid SSL Pinning:', event.errorDescription);
+            showingServerTrustAlert = true;
+            Alert.alert(
+                this.intl.formatMessage(messages.invlidePinningTitle),
```
Comment (enahum) [inline app/managers/network_manager.ts:163]:
```suggestion
                this.intl.formatMessage(messages.invalidePinningTitle),
```

**Comment by enahum** [comment]
I'll apply both suggestions in a batch once I'm back on desktop

File: app/client/websocket/index.ts [hunk 6]
```diff
@@ -190,6 +190,13 @@ export default class WebSocketClient {
             // reliable websockets are enabled this won't trigger a new sync.
             this.shouldSkipSync = false;
 
+            if (ev.message && typeof ev.message === 'object' && 'code' in ev.message && ev.message.code === 1015) {
+                // code 1015 means an error in the TLS handshake
```
Comment (enzowritescode) [inline app/client/websocket/index.ts:194]: Personal opinion - I feel like it's better to have a constant with a meaningful name than a comment to tell us

File: libraries/@mattermost/rnshare/android/src/main/java/com/mattermost/rnshare/ShareWorker.kt (hunk 3 above)
Comment (enzowritescode) [reply libraries/@mattermost/rnshare/android/src/main/java/com/mattermost/rnshare/ShareWorker.kt:61]: I was going to ask the same thing - especially considering I left comments on this code in the network client.

File: ios/Gekidou/Sources/Gekidou/PushNotification/PushNotification.swift [hunk 7]
```diff
@@ -75,9 +75,14 @@ public class PushNotification: NSObject {
                 url, withMethod: "POST", withBody: jsonData,
                 andHeaders: headers, forServerUrl: ackNotification.serverUrl) {[weak self] data, response, error in
                     if error != nil && ackNotification.isIdLoaded,
+                       let nsError = error as? NSError,
                        let fibonacciBackoffsInSeconds = self?.fibonacciBackoffsInSeconds,
                        let retryIndex = self?.retryIndex,
                        fibonacciBackoffsInSeconds.count > retryIndex {
+                        if nsError.code == -999 {
```
Comment (enzowritescode) [inline ios/Gekidou/Sources/Gekidou/PushNotification/PushNotification.swift:82]: Can you reference the constant so it's easier to read? i.e. no magic numbers

https://developer.apple.com/documentation/foundation/1508628-url_loading_system_error_codes/nsurlerrorcancelled

File: app/managers/network_manager.ts [hunk 8]
```diff
@@ -19,18 +20,41 @@ import {Client} from '@client/rest';
...
 const SERVER_CERTIFICATE_INVALID = -299;
+const SERVER_TRUST_EVALUATION_FAILED = -298;
+let showingServerTrustAlert = false;
+
+const messages = defineMessages({
+    invalidSslTitle: {
+        id: 'server.invalid.certificate.title',
+        defaultMessage: 'Invalid SSL certificate',
+    },
+    invalidSSLDescription: {
```
Comment (enzowritescode) [inline app/managers/network_manager.ts:40]: Inconsistent casing consistent between `invalidSslTitle` and `invalidSSLDescription` (one of them is SSL and the other is Ssl)

File: app/managers/network_manager.ts (hunk 5 above)
Comment (enzowritescode) [reply app/managers/network_manager.ts:163]: FWIW The above recommendation also has a typo. "invalide"

File: ios/Gekidou/Sources/Gekidou/Networking/Network+Delegate.swift [hunk 9]
```diff
@@ -1,24 +1,115 @@
...
-                                           persistence: URLCredential.Persistence.permanent)
+            try performDefaultValidation(trust)
+            
+            try performValidation(trust, forHost: host)
+            
+            try evaluate(trust, forHost: host, withCerts: certs)
+            
+            return (.useCredential, URLCredential(trust: trust), nil)
+        } catch {
+            return (.cancelAuthenticationChallenge, nil, error as? NetworkError)
```
Comment (enzowritescode) [inline ios/Gekidou/Sources/Gekidou/Networking/Network+Delegate.swift:69]: Is it worth logging anything here?

File: ios/Gekidou/Sources/Gekidou/Networking/Network.swift [hunk 10]
```diff
@@ -33,6 +37,39 @@ public class Network: NSObject {
         )
     }
     
+    internal func loadPinnedCertificates() {
+        if let certsPath = Bundle.app.resourceURL?.appendingPathComponent("certs") {
```
Comment (enzowritescode) [inline ios/Gekidou/Sources/Gekidou/Networking/Network.swift:41]: Minimize indentation, fail fast, and improve readability by:

```
guard let certsPath = Bundle.app.resourceURL?.appendingPathComponent("certs") else {
 	return
 }

// happy path
```

**Review by enzowritescode — CHANGES_REQUESTED** [review]

**Review by enzowritescode — APPROVED** [review]

**Comment by enahum** [comment]
sorry @enzowritescode wrong clicked the re-review button

**Review by esarafianou — APPROVED** [review]
@enahum feel free to merge

**Comment by cwarnermm** [comment]
@enahum - Is there product or developer documentation impact expected for this Engineering PR? Are you open to helping draft these updates?

**Comment by enahum** [comment]
> @enahum - Is there product or developer documentation impact expected for this Engineering PR? Are you open to helping draft these updates?

There was a PR alreqdy merged in the dev docs, no?

**Comment by cwarnermm** [comment]
@enahum - Yes, you are correct: https://github.com/mattermost/mattermost-developer-documentation/pull/1389. Thank you!

**Pull request merged**